pkg/controller/extension/reconciler.go

```diff
@@ -0,0 +1,248 @@
+	operationType := gardencorev1alpha1helper.ComputeOperationType(ex.ObjectMeta, ex.Status.LastOperation)
+	if err := r.updateStatusProcessing(ctx, ex, operationType, msg); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.actuator.Reconcile(ctx, ex); err != nil {
+		msg := "Unable to reconcile Extension resource"
+		r.updateStatusError(ctx, extensionscontroller.ReconcileErrCauseOrErr(err), ex, operationType, msg)
+		r.logger.Error(err, msg, "extension", ex.Name, "namespace", ex.Namespace)
+		return controllerutil.ReconcileErr(err)
+	}
+
+	msg = "Successfully reconciled Extension resource"
+	r.logger.Info(msg, "extension", ex.Name, "namespace", ex.Namespace)
+	if err := r.updateStatusSuccess(ctx, ex, operationType, msg); err != nil {
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *reconciler) delete(ctx context.Context, ex *extensionsv1alpha1.Extension) (reconcile.Result, error) {
+	hasFinalizer, err := controllerutil.HasFinalizer(ex, r.finalizerName)
+	if err != nil {
+		r.logger.Error(err, "Could not instantiate finalizer deletion")
+		return reconcile.Result{}, err
+	}
+
+	if !hasFinalizer {
+		r.logger.Info("Reconciling Extension resource causes a no-op as there is no finalizer.", "extension", ex.Name, "namespace", ex.Namespace)
+		return reconcile.Result{}, nil
+	}
+
+	operationType := gardencorev1alpha1helper.ComputeOperationType(ex.ObjectMeta, ex.Status.LastOperation)
+	if err := r.updateStatusProcessing(ctx, ex, operationType, "Deleting Extension resource."); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.actuator.Delete(ctx, ex); err != nil {
+		msg := "Error deleting Extension resource"
+		r.updateStatusError(ctx, extensionscontroller.ReconcileErrCauseOrErr(err), ex, operationType, msg)
+		r.logger.Error(err, msg, "extension", ex.Name, "namespace", ex.Namespace)
+		return controllerutil.ReconcileErr(err)
+	}
+
+	msg := "Successfully deleted Extension resource"
+	r.logger.Info(msg, "extension", ex.Name, "namespace", ex.Namespace)
+	if err := r.updateStatusSuccess(ctx, ex, operationType, msg); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	r.logger.Info("Extension resource deletion successful, removing finalizer.", "extension", ex.Name, "namespace", ex.Namespace)
+	if err := controllerutil.DeleteFinalizer(ctx, r.client, r.finalizerName, ex); err != nil {
+		r.logger.Error(err, "Error removing finalizer from Extension resource", "extension", ex.Name, "namespace", ex.Namespace)
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *reconciler) updateStatusProcessing(ctx context.Context, ex *extensionsv1alpha1.Extension, lastOperationType gardencorev1alpha1.LastOperationType, description string) error {
+	ex.Status.LastOperation = extensionscontroller.LastOperation(lastOperationType, gardencorev1alpha1.LastOperationStateProcessing, 1, description)
```

Comment:
I guess we want to use `TryUpdateStatus` here as well for all the status updates. The actuator might perform long-running operations.